digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/csvDataExporter/AssetReportCsvExporter.scala

```diff
@@ -0,0 +1,264 @@
+
+sealed trait SpecialAssetsTypeValues {
+  val id: Int
+}
+case object AllAssets extends SpecialAssetsTypeValues {val id = 1}
+case object PointAssets extends SpecialAssetsTypeValues {val id = 2}
+case object LinearAssets extends SpecialAssetsTypeValues {val id = 3}
+
+
+
+class AssetReportCsvExporter(roadLinkServiceImpl: RoadLinkService, eventBusImpl: DigiroadEventBus, userProviderImpl: UserProvider) extends CsvDataExporter(eventBusImpl: DigiroadEventBus){
+  override def withDynTransaction[T](f: => T): T = OracleDatabase.withDynTransaction(f)
+  override def withDynSession[T](f: => T): T = OracleDatabase.withDynSession(f)
+  val assetReporterDAO = new AssetReporterDAO()
+  val municipalityDao = new MunicipalityDao()
+  val userProvider: UserProvider = userProviderImpl
+
+  val mainHeaders = "Kunta;Tietolaji;kohteiden määrä (pistemäinen);Kohteita (viivamainen);Muokattu pvm (operaattori);Muokattu pvm (kuntakäyttäjä);Käyttäjä (operaattori);Käyttäjä (kuntakäyttäjä);*Liikennemerkit kpl lisätty 1.6.2020 jälkeen\r\n"
+
+
+  def exportAssetsByMunicipality( assetTypesList: List[Int], municipalitiesList: List[Int], extractPointsAsset: Boolean = false,
+                                  withTransaction: Boolean = true): List[ExportAssetReport] = {
+
+    def extractUserAndDateTime( assetReports: List[AssetReport], userNameList: Seq[String]): (String, Option[DateTime]) = {
+
+      val validUsers = assetReports.filter(asset => userNameList.contains(asset.modifiedBy))
+                                   .sortBy(_.modifiedDate.getMillis)(Ordering.Long.reverse)
+
+      if (validUsers.nonEmpty) {
+        val elem = validUsers.head
+        ( elem.modifiedBy , Some(elem.modifiedDate) )
+      }
+      else
+        ("", None)
+    }
+
+    def executeQuery(municipality: Int): List[AssetReport] = {
+      if (extractPointsAsset) {
+        assetReporterDAO.pointAssetQuery(Seq(municipality), assetTypesList)
+      }
+      else {
+        val linkIds = roadLinkServiceImpl.getVVHRoadLinksF(municipality)
+                                         .map(_.linkId)
+
+        assetReporterDAO.linearAssetQuery(linkIds, assetTypesList)
+      }
+    }
+
+    def extractData(municipality: Int): List[AssetReport] = {
+      if (withTransaction) {
+        withDynTransaction {
+          executeQuery(municipality)
+        }
+      }
+      else {
+        executeQuery(municipality)
+      }
+    }
+
+    def getUserAndMunicipalitiesInfo(): (Seq[User], List[MunicipalityInfo]) = {
+      if(withTransaction) {
+        withDynSession {
+          val userResults = userProvider.getUsers()
+          val municResults = municipalityDao.getMunicipalitiesNameAndIdByCode(municipalitiesList.toSet)
+
+          (userResults, municResults)
+        }
+      }
+      else{
+        val userResults = userProvider.getUsers()
+        val municResults = municipalityDao.getMunicipalitiesNameAndIdByCode(municipalitiesList.toSet)
+
+        (userResults, municResults)
+      }
```

Comment:
Move this code to a function since is repeated
    
```scala
 val userResults = userProvider.getUsers()
        val municResults = municipalityDao.getMunicipalitiesNameAndIdByCode(municipalitiesList.toSet)

        (userResults, municResults)
```

Reply:
Function created to avoid repeated code